### PR TITLE
add support for streaming in websockets && fix issue #2210 && allow t…

### PR DESCRIPTION
**Pull request by romankoblov**
…o run streaming queries in batch run concurrently

### CODE REVIEW

File: rpc/handler.go [hunk 1]
```diff
@@ -73,6 +71,13 @@ type handler struct {
 	maxBatchConcurrency uint
 }
 
+// Custom bytes type which writes to JSON as is
+type JSONBytes []byte
```
Comment (AskAlexSharov) [inline rpc/handler.go:75]: Native json.RawMessage can be used here?

Comment (AskAlexSharov) [reply rpc/handler.go:75]: I will review tomorrow. Late here.

**Comment by AskAlexSharov** [comment]
Will review tomorrow, late here

File: rpc/handler.go (hunk 1 above)
Comment (romankoblov) [reply rpc/handler.go]: Fixed

**Comment by romankoblov** [comment]
> Will review tomorrow, late here

Sure, please check that its ok, I don't know go, so it is probably very bad code, but it was more like proof of concept

**Comment by romankoblov** [comment]
I've simplified it a bit and now batch requests are streamable too. But there is some issues with tests and I cannot understand whats going on, can you give me some pointers?

**Comment by AskAlexSharov** [comment]
cool. 

- I don't understand 1 place: there are 2 `for` loops, can we through away second? and do all the work in first?
- let's move mutex lock/unlock inside `writeToStream` func.

Need to test next cases: 
- 1 batch has [streaming, non-streaming, streaming] - works
- 1 batch has [streaming, failed-non-streaming, streaming] - doesn't work (for example pass invalid block hash: "0xRR") - it's because streamable handleCallMsg also return's error.
- 1 batch has [failed-streaming, non-streaming, streaming] - doesn't work (for example pass invalid block hash: "0xRR")

To support WS already good-enough target.

In future (not in this PR) - as I understand - need to get read from `h.conn.writeJSON` and replace it by stream(or something similar) - to allow send to client responses when they are ready (because order doesn't require).

**Comment by romankoblov** [comment]
> cool.
> 
> * I don't understand 1 place: there are 2 `for` loops, can we through away second? and do all the work in first?

Fixed, but now subscribe queries is forced to be non-streaming. I cannot find them in current code, but maybe there is something in plans.

> * let's move mutex lock/unlock inside `writeToStream` func.
> 

Fixed.

> Need to test next cases:
> 
> * 1 batch has [streaming, non-streaming, streaming] - works
> * 1 batch has [streaming, failed-non-streaming, streaming] - doesn't work (for example pass invalid block hash: "0xRR") - it's because streamable handleCallMsg also return's error.
> * 1 batch has [failed-streaming, non-streaming, streaming] - doesn't work (for example pass invalid block hash: "0xRR")
> 

Fixed.

> To support WS already good-enough target.
> 
> In future (not in this PR) - as I understand - need to get read from `h.conn.writeJSON` and replace it by stream(or something similar) - to allow send to client responses when they are ready (because order doesn't require).

You mean streaming for websockets? Or just to avoid memory copy and double buffering between jsoniter buffer to socket buffer? 
If first one, then problem that streaming requires support from both sides, otherwise big queries will buffer on client side and then crash client (instead of server). For http it is possible to do even in browser with custom json parser and XHR requests, but for websockets I'm not aware of any libraries which support streaming processing of websocket frames. 

Better to do something like:

```
for block in 0...N {
   if len(stream.Buffer()) > 16*MB {
        stream.write("paginate: true"); // just to avoid client request same query multiple times with different ranges to check if query was paginated
        break;
   }
   process(block);
}
```

**Comment by AskAlexSharov** [comment]
I mean streaming of batches for non-websocket. (yes, ws - is different animal).
"support from both sides" - yes, but now clients have chance to do so. (we don't throw to Client much, only what client asked :-) )

 
I will review later today.

**Comment by romankoblov** [comment]
> I mean streaming of batches for non-websocket. (yes, ws - is different animal).

For http there will be stream in arguments (instead of nil), so batch call will just use it, instead of writing via h.conn.writejSON

> "support from both sides" - yes, but now clients have chance to do so. (we don't throw to Client much, only what client asked :-) )

But client doesn’t know how much data there will be and not always have control of receiving buffers, server on other end can do something if request is too big

> 
> I will review later today.

BTW, for http it is possible to do this:
all non-streaming requests will stream when done, streaming requests will try to acquire mutex before streaming, that way there still be parallel execution of non-streaming requests, but streaming requests will be processed sequential without intermediate buffer. Worth trying?

**Comment by AskAlexSharov** [comment]
```
all non-streaming requests will stream when done, streaming requests will try to acquire mutex before streaming, that way there still be parallel execution of non-streaming requests, but streaming requests will be processed sequential without intermediate buffer. Worth trying?
```
I think - what we have now (on batch streamable request - use intermediate buffer and send it to client when sub-request is done) - is enough. And better than make requests sequential. Thank you.

**Comment by romankoblov** [comment]
> ```
> all non-streaming requests will stream when done, streaming requests will try to acquire mutex before streaming, that way there still be parallel execution of non-streaming requests, but streaming requests will be processed sequential without intermediate buffer. Worth trying?
> ```
> 
> 
>      
>   
> 
> I think - what we have now (on batch streamable request - use intermediate buffer and send it to client when sub-request is done) - is enough. And better than make requests sequential. Thank you.

Thanks for help with PR!

**Pull request merged**